contrib/inventory/foreman.ini

```diff
@@ -108,15 +166,77 @@
 #
 # would put the host into the additional Ansible groups:
 #
-#     - myapp-webtier-datacenter1
-#     - myapp-webtier
-#     - myapp
-#     - mysubnet-image
+#     - myApp_web_tier_DataCenter_1
+#     - myApp_web_tier
+#     - myApp
+#     - mysubnet_image
+#
+# by recursively resolving the hostgroups, getting the parameter keys and
+# and values and doing a Python *string.format()* like replacement on it.
+#
+# If you want host group names with a prefix similar to "foreman_hostgroup"
+# or "foreman_hostcollection" for your parameter-derived group names, include the
+# prefix in your group_patterns. Say your hosts have a parameter "sm_customer"
+# with values "dba", "middleware", and "finance". Then
+# 
+#     group_patterns = ["foreman/param/sm_customer/{sm_customer}"]
```

Comment:
It's probably more than a single line though if a user has to provide group_patterns? At the least, it appears that `group_patterns` duplicates functionality. Have you tried anything like the following instead of `group_patterns` and `want_generated_parents`?

cat constructed_inventory.yml
```
plugin: constructed
keyed_groups:
  - key: sm_customer
    prefix: foreman_param_sm_customer
    parent_groups:
      - foreman_param_smCustomer
      - foreman_param
      - foreman
```

ansible-inventory -i contrib/inventory/foreman.py -i constructed_inventory.yml --graph

You can also makes groups of groups via `groups` in a constructed config to run after your keyed_groups have been created:
```
plugin: constructed
groups:
  foreman: group_names | select('match', 'foreman_*') | list | length >= 1
```

Note that the contrib scripts aren't being actively developed since inventory plugins are available now. Have you checked out the foreman inventory plugin https://docs.ansible.com/ansible/latest/plugins/inventory/foreman.html? Features should be added there instead.

Reply:
`group_patterns` is existing functionality in the foreman.py inventory script. I was merely correcting/completing the documentation for it in light of these other features. (Some pains were taken to preserve backwards compatibility with those features, btw. If you don't change your foreman.ini file and just replace the script, you get no surprises.)

Foreman group names map poorly and ambiguously to Ansible group names because of the different sets of characters allowed in each. The two techniques you show above require (a) knowing in advance -- and replicating -- the group hierarchy from foreman, or (b) pattern matching against strings which have already lost crucial hierarchical information by the time the plugin config takes affect.

Given that the contrib/README.md still says "contributions welcome Send in pull requests to add scripts of your own. The sky is the limit!" and that there are pending pull requests from others, the news that "contrib scripts aren't being actively developed" is a little unsettling. Are you suggesting that we should port the functionality we need to the foreman inventory plugin rather than share those already developed features with people using the inventory script? Perhaps that is the way to go long-term, but this is functionality that we've needed yesterday. I see nothing in https://docs.ansible.com/ansible/latest/plugins/inventory/foreman.html that does anything wrt case of the group names or parameter values brought over from foreman, or any mechanism for preserving foreman's group hierarchy, or any mechanism for setting group_vars, particularly `ansible_group_priority` which must be done in the inventory to work, all of which is possible in our patched foreman.py. Granted, the foreman inventory plugin is news to me, so there's surely more to it (and inventory plugins in general) than I'm aware of. But I don't think that's reason enough to leave otherwise functioning inventory scripts to rot from neglect.

Reply:
Ah, my mistake about group_patterns. It looked like you were adding more behavior on top of that option. Regarding a), `parent_group` names are templated and it appeared to be what the script does, but I may be over simplifying your problem or misreading the code.

The documentation should be updated. I wouldn't say the scripts are “rotting”… tests and bug fixes are still merged. Inventory plugins are shipped with Ansible and share code, so they have benefits over the scripts. Since the plugins are expected to be able to recreate the output of the scripts if we add new features to the scripts we stray further from parity. This also seems like the sort of thing that has value for inventory in general rather than a specific source. Anyway, just my opinion.

Reply:
The notion of "the sort of thing that has value for inventory in general rather than a specific source" resonates very well with me. Having now had a chance to look (briefly) over the foreman inventory plugin and the inventory plugin supporting code, several things seem to need addressing.

1. Preserving Parent Groups: Preserving host group hierarchies for sources that have them (like foreman) is crucial. Because of necessary group name mangling (to_safe_group_name() and .lower()) it's impossible to reconstruct group hierarchies from "leaf group" names after the fact.

2. Group Vars: I see support for setting host vars (`compose`), but general support for setting group vars seems missing. In particular, `ansible_group_priority` has to work at the inventory group level.

3. More flexibility in group name mangling: It's insufficient to ensure group names are safe for Ansible (again, think to_safe_group_name() and .lower()); they also must meet the goals and standards of the organization.

I don't see how to solve these problems using only `constructed` plugin's `compose`, `groups`, and `keyed_groups` options. They certainly _can_ be addressed in individual inventory scripts (as the proposed patch does for foreman) or in inventory plugins on a case by case basis. For example, porting my changes to the foreman inventory plugin should be pretty straightforward, and I'm willing to do that if you think there's a chance the patches might be accepted. However, addressing these issues more generally at the inventory plugin infrastructure level is beyond the scope of this PR.